07_StagingAndMultipleQueues/main.cpp

```diff
@@ -2,78 +2,580 @@
+			// constexpr uint32_t FAMILY_INDICES_CNT = 3; // TODO: test on intel integrated GPU (which allows only one queue family)
+			std::array familyIndices = { getTransferUpQueue()->getFamilyIndex(), getComputeQueue()->getFamilyIndex() };
+			imgParams.queueFamilyIndexCount = familyIndices.size();
+			imgParams.queueFamilyIndices = familyIndices.data();
+			imgParams.preinitialized = false;
+
+			images[imageIdx] = m_device->createImage(std::move(imgParams));
+			auto imageAllocation = m_device->allocate(images[imageIdx]->getMemoryReqs(), images[imageIdx].get(), IDeviceMemoryAllocation::EMAF_NONE);
+
+			if (!imageAllocation.isValid())
+				logFailAndTerminate("Failed to allocate Device Memory compatible with our GPU Buffer!\n");
+
+			
+			IGPUCommandBuffer::SImageMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier> imageLayoutTransitionBarrier0; // TODO: better names maybe?
+			IGPUCommandBuffer::SImageMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier> imageLayoutTransitionBarrier1;
+			{
+				IImage::SSubresourceRange imgSubresourceRange{};
+				imgSubresourceRange.aspectMask = IImage::E_ASPECT_FLAGS::EAF_COLOR_BIT;
+				imgSubresourceRange.baseMipLevel = 0u;
+				imgSubresourceRange.baseArrayLayer = 0u;
+				imgSubresourceRange.levelCount = 1;
+				imgSubresourceRange.layerCount = 1u;
+
+				imageLayoutTransitionBarrier0.barrier.dep.srcAccessMask = ACCESS_FLAGS::NONE;
+				imageLayoutTransitionBarrier0.barrier.dep.dstAccessMask = ACCESS_FLAGS::MEMORY_WRITE_BITS;
+				imageLayoutTransitionBarrier0.barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::NONE;
+				imageLayoutTransitionBarrier0.barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS; //TODO: COPY_BIT?
+				imageLayoutTransitionBarrier0.oldLayout = asset::IImage::LAYOUT::UNDEFINED;
+				imageLayoutTransitionBarrier0.newLayout = asset::IImage::LAYOUT::TRANSFER_DST_OPTIMAL; // TODO: use more suitable layout
+				imageLayoutTransitionBarrier0.image = images[imageIdx].get();
+				imageLayoutTransitionBarrier0.subresourceRange = imgSubresourceRange;
+
+				imageLayoutTransitionBarrier1.barrier.dep.srcAccessMask = ACCESS_FLAGS::MEMORY_WRITE_BITS; 
+				imageLayoutTransitionBarrier1.barrier.dep.dstAccessMask = ACCESS_FLAGS::NONE;
+				imageLayoutTransitionBarrier1.barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS; //TODO: COPY_BIT?
+				imageLayoutTransitionBarrier1.barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::NONE;
+				imageLayoutTransitionBarrier1.oldLayout = asset::IImage::LAYOUT::TRANSFER_DST_OPTIMAL;
+				imageLayoutTransitionBarrier1.newLayout = asset::IImage::LAYOUT::GENERAL; // TODO: use more suitable layout
+				imageLayoutTransitionBarrier1.image = images[imageIdx].get();
+				imageLayoutTransitionBarrier1.subresourceRange = imgSubresourceRange;
+			}
+
+			core::smart_refctd_ptr<ISemaphore> imgFillSemaphore = m_device->createSemaphore(0); // TODO: don't create semaphore every iteration
```

Comment:
you can create one outside the loop and just reuse it, you need to keep track of the next signal value as well